File: java/Tutorials/BiometricStandards/C/NTemplateToANTemplate/NTemplateToANTemplate.c
#include <TutorialUtils.h>

#ifdef N_MAC_OSX_FRAMEWORKS
	#include <NCore/NCore.h>
	#include <NBiometricClient/NBiometricClient.h>
	#include <NBiometrics/NBiometrics.h>
	#include <NMedia/NMedia.h>
	#include <NLicensing/NLicensing.h>
#else
	#include <NCore.h>
	#include <NBiometricClient.h>
	#include <NBiometrics.h>
	#include <NMedia.h>
	#include <NLicensing.h>
#endif

const NChar title[] = N_T("NTemplateToANTemplate");
const NChar description[] = N_T("Demonstrates creation of ANTemplate from NTemplate");
const NChar version[] = N_T("9.0.0.0");
const NChar copyright[] = N_T("Copyright (C) 2009-2017 Neurotechnology");

int usage()
{
	printf(N_T("usage:\n"));
	printf(N_T("\t%s [NTemplate] [ANTemplate] [Tot] [Dai] [Ori] [Tcn]\n"), title);
	printf(N_T("\n"));
	printf(N_T("\tNTemplate   - filename with NTemplate\n"));
	printf(N_T("\tANTemplate  - filename for NTemplate\n"));
	printf(N_T("\tTot - specifies type of transaction\n"));
	printf(N_T("\tDai - specifies destination agency identifier\n"));
	printf(N_T("\tOri - specifies originating agency identifier\n"));
	printf(N_T("\tTcn - specifies transaction control number\n"));

	return 1;
}

NResult NTemplateToANTemplateFile(
	HNBuffer hTemplateBuffer, // buffer that contains packed NTemplate
	NChar * szFileName, // pointer to string that specifies filename to store ANTemplate
	const NChar * szTot, // pointer to tring that specifies type of transaction
	const NChar * szDai, // pointer to string that specifies destination agency identifier
	const NChar * szOri, // pointer to string that specifies originating agency identifier
	const NChar * szTcn // pointer to string that specifies transaction control number
)
{
	HNTemplate hNTemplate = NULL; // handle to NTemplate object
	HANTemplate hANTemplate = NULL; // handle to ANTemplate object
	NResult result;

	// Create NTemplate object from packed NTemplate
	result = NTemplateCreateFromMemoryN(hTemplateBuffer, 0, NULL, &hNTemplate);
	if(NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("error in NTemplateCreateFromMemory, error code: %d\n"), result);
		goto FINALLY;
	}

	// Create ANTemplate object from NFRecord object
	result = ANTemplateCreateFromNTemplateEx(AN_TEMPLATE_VERSION_CURRENT, szTot, szDai, szOri, szTcn, NTrue, hNTemplate, N_OBJECT_REF_RET, &hANTemplate);
	if(NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("error in ANTemplateCreateFromNTemplate, error code: %d\n"), result);
		goto FINALLY;
	}

	// Store ANTemplate object in file
	result = ANTemplateSaveToFile(hANTemplate, szFileName, 0);
	if(NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("error in ANTemplateSaveToFile, error code: %d\n"), result);
		goto FINALLY;
	}

	result = N_OK;
FINALLY:
	{
		NResult result2 = NObjectSet(NULL, &hNTemplate);
		if NFailed((result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
		result2 = NObjectSet(NULL, &hANTemplate);
		if NFailed((result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
	}

	return result;
}

int main(int argc, NChar **argv)
{
	NResult result = N_OK;
	HNBuffer hBuffer = NULL;
	// Depending on NTemplate contents choose the licenses: if you will have only finger templates in NTemplate - leave finger templates license only.
	const NChar * components = N_T("Biometrics.Standards.FingerTemplates,Biometrics.Standards.PalmTemplates");
	NBool available = NFalse;

	OnStart(title, description, version, copyright, argc, argv);

	if(argc < 7)
	{
		OnExit();
		return usage();
	}

	// check the license first
	result = NLicenseObtainComponents(N_T("/local"), N_T("5000"), components, &available);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NLicenseObtainComponents() failed, result = %d\n"), result);
		goto FINALLY;
	}

	if (!available)
	{
		printf(N_T("Licenses for %s not available\n"), components);
		result = N_E_NOT_ACTIVATED;
		goto FINALLY;
	}

	result = NFileReadAllBytesCN(argv[1], &hBuffer);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NFileReadAllBytesN() failed, result = %d\n"), result);
		goto FINALLY;
	}
	
	if ((strlen(argv[3]) < 3) || (strlen(argv[3]) > 4))
	{
		printf(N_T("tot parameter should be 3 or 4 characters length\n"));
		result = N_E_FAILED;
		goto FINALLY;
	}

	result = NTemplateToANTemplateFile(hBuffer, argv[2], argv[3], argv[4], argv[5], argv[6]);
	if(NFailed(result))
	{
		goto FINALLY;
	}

	result = N_OK;
FINALLY:
	{
		NResult result2 = NObjectSet(NULL, &hBuffer);
		if NFailed((result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
		result2 = NLicenseReleaseComponents(components);
		if (NFailed(result2)) PrintErrorMsg(N_T("NLicenseReleaseComponents() failed, result = %d\n"), result2);
	}

	OnExit();
	return result;
}

File: java/Tutorials/BiometricStandards/Java/ntemplate-to-antemplate/src/main/java/com/neurotec/tutorials/NTemplateToANTemplate.java
package com.neurotec.tutorials;

import java.io.IOException;

import com.neurotec.biometrics.NTemplate;
import com.neurotec.biometrics.standards.ANTemplate;
import com.neurotec.io.NBuffer;
import com.neurotec.io.NFile;
import com.neurotec.licensing.NLicense;
import com.neurotec.tutorials.util.LibraryManager;
import com.neurotec.tutorials.util.Utils;

public final class NTemplateToANTemplate {

	private static final String DESCRIPTION = "Converting NTemplate to ANTemplate";
	private static final String NAME = "ntemplate-to-antemplate";
	private static final String VERSION = "9.0.0.0";

	private static void usage() {
		System.out.println("usage:");
		System.out.format("\t%s [NTemplate] [ANTemplate] [Tot] [Dai] [Ori] [Tcn]%n", NAME);
		System.out.println("");
		System.out.println("\t[NTemplate]     - filename of NTemplate.");
		System.out.println("\t[ANTemplate]    - filename of ANTemplate.");
		System.out.println("\t[Tot] - specifies type of transaction.");
		System.out.println("\t[Dai] - specifies destination agency identifier.");
		System.out.println("\t[Ori] - specifies originating agency identifier.");
		System.out.println("\t[Tcn] - specifies transaction control number.");
		System.out.println("");
	}

	public static void main(String[] args) {
		LibraryManager.initLibraryPath();

		// Depending on NTemplate contents choose licenses: if you only have finger templates in NTemplate, leave finger templates license only.
		final String components = "Biometrics.Standards.FingerTemplates,Biometrics.Standards.PalmTemplates";

		Utils.printTutorialHeader(DESCRIPTION, NAME, VERSION, args);

		if (args.length < 6) {
			usage();
			System.exit(1);
		}

		if (args[0] == "/?" || args[0] == "help") {
			usage();
			System.exit(-1);
		}

		NTemplate nTemplate = null;
		ANTemplate anTemplate = null;

		try {
			String nTemplateFileName = args[0];

			String tot = args[2]; // type of transaction
			String dai = args[3]; // destination agency identifier
			String ori = args[4]; // originating agency identifier
			String tcn = args[5]; // transaction control number

			if ((tot.length() < 3) || (tot.length() > 4)) {
				System.out.println("Tot parameter should be 3 or 4 characters length.");
				System.exit(-1);
			}

			NBuffer packedNTemplate = NFile.readAllBytes(nTemplateFileName);

			// Creating NTemplate object from packed NTemplate
			nTemplate = new NTemplate(packedNTemplate);

			if (!NLicense.obtainComponents("/local", 5000, components)) {
				System.err.println("Could not obtain licenses for components: " + components);
				System.exit(-1);
			}

			// Creating ANTemplate object from NTemplate object
			anTemplate = new ANTemplate(ANTemplate.VERSION_CURRENT, tot, dai, ori, tcn, true, nTemplate);

			// Storing ANTemplate object in file
			anTemplate.save(args[1]);
			System.out.println("Program produced file: " + args[1]);
		} catch (Throwable th) {
			Utils.handleError(th);
		} finally {
			try {
				NLicense.releaseComponents(components);
			} catch (IOException e) {
				e.printStackTrace();
			}
			if (nTemplate != null) nTemplate.dispose();
			if (anTemplate != null) anTemplate.dispose();
		}
	}
}
